### PR TITLE
[MM-26812] Add support for resumable file uploads

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR adds the required API endpoints and app logic to support resumable file uploads.

#### Ticket

https://mattermost.atlassian.net/browse/MM-26812

#### Related PRs

https://github.com/mattermost/mattermost-api-reference/pull/566

### CODE REVIEW

File: model/client4.go
Comment (isacikgoz) [inline model/client4.go:321]:
```suggestion
	return fmt.Sprintf("%s/%s", c.GetUploadsRoute(), uploadId)
```

File: model/upload_session.go
Comment (isacikgoz) [inline model/upload_session.go:14]: By resource you mean session right? If so, maybe we can just say `session`.

File: api4/user.go
Comment (streamer45) [reply api4/user.go:2542]: That's a PM question I feel. I've purposely decided to start with the least possible permissions.
We might even consider `PERMISSION_EDIT_OTHER_USERS` if it makes sense but I'd defer to @esethna on this one.

File: app/upload.go
Comment (streamer45) [reply app/upload.go:44]: The idea is that before doing any DB call we check if the session struct is valid.

**Comment by esethna** [comment]
Adding @enahum for the mobile perspective on this in flaky networks

**Comment by enahum** [comment]
@esethna what do you want me to review?

File: api4/upload.go
Comment (enahum) [inline api4/upload.go:16]: why new endpoints?

**Comment by esethna** [comment]
@enahum just verifying that this would work to solve the issues we see on mobile where file uploads have to be restarted because of a temporary network disconnect/hiccup

**Comment by enahum** [comment]
I’m afraid I don’t know anything about that but I was expecting to use the same endpoints without having to make any changes on the client.

Maybe someone needs to explain how this works and we can discuss the approach

**Comment by esethna** [comment]
@enahum you filed the related ticket here haha: https://mattermost.atlassian.net/browse/MM-22922

**Comment by enahum** [comment]
@esethna I mean from the server points of view

File: api4/upload.go
Comment (streamer45) [reply api4/upload.go:16]: For a couple of reasons mainly:

1. To support this feature I've introduced a new `UploadSession` entity  which is separate from `FileInfo` and it made sense for it to have its own endpoint.
2. The `/files` endpoint (POST method) is already polluted with two different ways of uploading files (multipart and file in body). Moreover, both suffer from a problem we are trying to eliminate which is the loading of whole files in memory. Since the required changes are non trivial and possibly breaking, I've thought that implementing a totally new codepath would be the fastest, cleanest and most reliable way to both solve this issue and at the same time implement the requested feature. The idea is that once this has been thoroughly tested and successfully implemented on the front-end we can gradually move away from the previous calls through deprecation.

**Comment by streamer45** [comment]
> I’m afraid I don’t know anything about that but I was expecting to use the same endpoints without having to make any changes on the client.
> 
> Maybe someone needs to explain how this works and we can discuss the approach

This is (briefly) how it works:

1. To initialize a new file upload the user creates an `UploadSession`. (`POST /uploads`)
2. To upload the file the user sends the content of the file. (`POST /uploads/{upload_id}`)

Now there are two outcomes:

- All data was sent. The upload completes and a `FileInfo` is returned.
- Not all data was sent. The upload is incomplete. The user should fetch the upload session (`GET /uploads/{upload_id}`) look for the `FileOffset` field (which indicates how much data the server has received) and send data starting from that value. Repeat this process until all data is sent and the upload completes.

More details are available in the docs PR:
https://github.com/mattermost/mattermost-api-reference/pull/566

**Comment by enahum** [comment]
@streamer45 

> This is (briefly) how it works:
To initialize a new file upload the user creates an UploadSession. (POST /uploads)
To upload the file the user sends the content of the file. (POST /uploads/{upload_id})

So instead of just one request to upload a file we now need two? I don't really get the need for that as well as adding a lot of complexity o the client as now we need to keep track of an ID being returned by the server. Also the fact that every request can be a point of failure, specially on a mobile network.

Would it be better if the client sends what you call the `upload_id` the same way we do with posts `pending_post_id`? or what about a checksum, either in the body or in the header of the upload request?

**Comment by streamer45** [comment]
> So instead of just one request to upload a file we now need two? I don't really get the need for that as well as adding a lot of complexity o the client as now we need to keep track of an ID being returned by the server. Also the fact that every request can be a point of failure, specially on a mobile network.

I think the complexity is minimized as much as possible on the client. I've tried to follow industry standards on this one.
Every other public implementation I could find more or less worked this way. Let's not forget we are introducing a new feature that implicitly demands more logic cause of state handling involved.

> Would it be better if the client sends what you call the `upload_id` the same way we do with posts `pending_post_id`? 

I can see how sending a client generated `upload_id` can save us the initial upload session creation request but we would still have to query the server in case of failure to know how much data (if any) was received.

From an API design perspective I am not particularly in love with the idea of accepting a client provided id and essentially combine two rather different operations just to save an HTTP call. 

Also, I am not sure I understand how the ID tracking complexity is going to change whether the ID is client generated vs server generated. I ask for your input on this cause I might be lacking some client-side vision.

> or what about a checksum, either in the body or in the header of the upload request?

Were you thinking of using the checksum as a way to keep track of the file being uploaded?
It could certainly be added but I fear it might introduce a significant performance penalty on the client, especially for big files so I wouldn't make it a requirement.

**Comment by enahum** [comment]
@streamer45 I see, I was unaware there is a standard or protocol for this. Exploring the internet I found this https://tus.io/protocols/resumable-upload.html but it differs in implementation.

Also can you add a usage example? Thanks

**Comment by streamer45** [comment]
> @streamer45 I see, I was unaware there is a standard or protocol for this. Exploring the internet I found this https://tus.io/protocols/resumable-upload.html but it differs in implementation.

There isn't a standard protocol but there are similarities across the major public implementations. They all share the concept of an upload session which is created prior to the actual uploading happening.
The `tus` protocol was considered but in the end I opted for a custom and simpler solution since we have more logic behind the scenes and we don't really need to provide a general purpose file upload service.

Some examples I used as guidelines:

- https://developers.facebook.com/docs/graph-api/resumable-upload-api/
- https://developers.google.com/drive/api/v3/manage-uploads#resumable

> Also can you add a usage example? Thanks

```go
package main

import (
	"fmt"
	"io"
	"os"

	"github.com/mattermost/mattermost-server/v5/model"
)

func main() {
	// login
	client := model.NewAPIv4Client("http://localhost:8065")
	_, resp := client.Login("username", "password")
	if resp.Error != nil {
		fmt.Fprintf(os.Stderr, "%v\n", resp.Error)
		return
	}

	// open file
	file, err := os.Open("upload")
	if err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		return
	}

	// get file size
	info, err := file.Stat()
	if err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		return
	}

	// create upload session
	us, resp := client.CreateUpload(&model.UploadSession{
		ChannelId: "4i6jn8r483nnuqnibnmgz8jo4o",
		Filename:  info.Name(),
		FileSize:  info.Size(),
	})
	if resp.Error != nil {
		fmt.Fprintf(os.Stderr, "%v\n", resp.Error)
		return
	}

	/*
		// if the upload session creation were to fail for some network error we have two options:
		// - Create a brand new upload session. I would suggest this since no data
		// has been sent yet.
		// - Query the server to figure out if the session was created. (this is
		// needed to possibly resume uploads across different clients), e.g.:
		uss, resp := client.GetUploadsForUser(user.Id)
		if resp.Error != nil {
			fmt.Fprintf(os.Stderr, "%v\n", resp.Error)
			return
		}
		// this is where I think sending the checksum could help. We could
		// identify the file more easily and more accurately.
		for _, us := range uss {
			if us.Filename == info.Name() && us.FileSize == info.Size() {
				// upload session found, proceeed with uploading logic
				break
			}
		}
	*/

	// We simulate an incomplete upload by sending only the first 5MBs.'
        // This is actually an implementation requirement. First part of a resumable upload 
        // has to be at least 5MBs in size.
	rd := &io.LimitedReader{
		R: file,
		N: 5 * 1024 * 1024,
	}

	// upload data
	var finfo *model.FileInfo
	finfo, resp = client.UploadData(us.Id, rd)
	if resp.Error != nil {
		fmt.Fprintf(os.Stderr, "%v\n", resp.Error)
		return
	}

	for finfo == nil {
		// query the server to figure out how much data it received.
		us, resp = client.GetUpload(us.Id)
		if resp.Error != nil {
			fmt.Fprintf(os.Stderr, "%v\n", resp.Error)
			return
		}

		// we seek to the returned offset and send the remaining bytes.
		file.Seek(us.FileOffset, io.SeekStart)
		// send the file in chunks of 64KBs to simulate a very spotty connection.
		rd.N = 32 * 1024
		finfo, resp = client.UploadData(us.Id, rd)
		if resp.Error != nil {
			fmt.Fprintf(os.Stderr, "%v\n", resp.Error)
			return
		}
	}

	fmt.Printf("Upload done: %+v\n", finfo)
}
```

**Comment by enahum** [comment]
Thanks for the go example using the go client but I was thinking more along the lines of a REST example.

Am I to assume that this is an octet stream upload?

**Comment by streamer45** [comment]
> Thanks for the go example using the go client but I was thinking more along the lines of a REST example.
> 
> Am I to assume that this is an octet stream upload?

It's a simple upload with the bytes content in the body (no multipart).  
For example, curl sends the data by default as `application/x-www-form-urlencoded`.
The content-type is not really considered. The server currently guesses it from the filename extension.

As for a lower level example, I can provide one with CURL:

#### 1. Create upload session:
```sh
curl -v 'http://localhost:8065/api/v4/uploads' -H 'Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw' -H 'Content-Type: application/json' --data-binary '{"channel_id": "4i6jn8r483nnuqnibnmgz8jo4o", "filename": "upload", "file_size": 7082489}'
```

##### Request:
```
> POST /api/v4/uploads HTTP/1.1
> Host: localhost:8065
> User-Agent: curl/7.71.1
> Accept: */*
> Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw
> Content-Type: application/json
> Content-Length: 88
```

##### Response:
```
< HTTP/1.1 201 Created
< Content-Type: application/json
< Vary: Accept-Encoding
< X-Request-Id: 3j1qiqio6brd8bnegghiukhhxr
< X-Version-Id: 5.28.0..ea71bf8023fe529a2125d71e6f29fa4b.true
< Date: Mon, 17 Aug 2020 11:50:53 GMT
< Content-Length: 198               
```

##### Response body:
```json
{"id":"xae3k3emifyruywe47crigjinw","create_at":1597665053893,"user_id":"fc6suoon9pbbpmhrb9c967paxe","channel_id":"4i6jn8r483nnuqnibnmgz8jo4o","filename":"upload","file_size":7082489,"file_offset":0}
```

#### 2. Upload data

```sh
curl -v 'http://localhost:8065/api/v4/uploads/yzn1pj4z1bynirzi46zuiodguh' -H 'Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw' --data-binary @upload
```

##### Request
```
> POST /api/v4/uploads/yzn1pj4z1bynirzi46zuiodguh HTTP/1.1
> Host: localhost:8065
> User-Agent: curl/7.71.1
> Accept: */*
> Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw
> Content-Length: 7082489
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue

< HTTP/1.1 100 Continue
```

##### Response:

```
< HTTP/1.1 200 OK
< Content-Type: application/json
< Vary: Accept-Encoding
< X-Request-Id: ba97ddrpnb8turi1q148qr7sja
< X-Version-Id: 5.28.0..ea71bf8023fe529a2125d71e6f29fa4b.true
< Date: Mon, 17 Aug 2020 11:54:08 GMT
< Content-Length: 201
```

##### Response body:

```json
{"id":"c44uh47yki8zje578yxj7wgcgc","user_id":"fc6suoon9pbbpmhrb9c967paxe","create_at":1597665248548,"update_at":1597665248548,"delete_at":0,"name":"upload","extension":"","size":7082489,"mime_type":""}
```

To simulate an incomplete upload you can add `--limit-rate 1m` to curl and stop the command before it's finished:

#### 1. Create upload session
```sh
curl -v 'http://localhost:8065/api/v4/uploads' -H 'Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw' -H 'Content-Type: application/json' --data-binary '{"channel_id": "4i6jn8r483nnuqnibnmgz8jo4o", "filename": "upload", "file_size": 7082489}'
```

##### Request
```
> POST /api/v4/uploads HTTP/1.1
> Host: localhost:8065
> User-Agent: curl/7.71.1
> Accept: */*
> Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw
> Content-Type: application/json
> Content-Length: 88
```

##### Response
```
< HTTP/1.1 201 Created
< Content-Type: application/json
< Vary: Accept-Encoding
< X-Request-Id: rozjtb59ft8nunynufy91ns81h
< X-Version-Id: 5.28.0..ea71bf8023fe529a2125d71e6f29fa4b.true
< Date: Mon, 17 Aug 2020 12:00:37 GMT
< Content-Length: 198
```

##### Response body
```json
{"id":"t8edyq9t33d3zmz9cgwgfiynhc","create_at":1597665637367,"user_id":"fc6suoon9pbbpmhrb9c967paxe","channel_id":"4i6jn8r483nnuqnibnmgz8jo4o","filename":"upload","file_size":7082489,"file_offset":0}                                                                                                                                                                                               
```


#### 2. Upload data
```
curl -v 'http://localhost:8065/api/v4/uploads/t8edyq9t33d3zmz9cgwgfiynhc' -H 'Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw' --data-binary @upload --limit-rate 1m
```

```
> POST /api/v4/uploads/t8edyq9t33d3zmz9cgwgfiynhc HTTP/1.1
> Host: localhost:8065
> User-Agent: curl/7.71.1
> Accept: */*
> Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw
> Content-Length: 7082489
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue

< HTTP/1.1 100 Continue
```

#### 3. Query server to figure out how much data was received

```
curl -v 'http://localhost:8065/api/v4/uploads/t8edyq9t33d3zmz9cgwgfiynhc' -H 'Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw'
```

##### Request
                                                                                                                                                                                            
```
> GET /api/v4/uploads/t8edyq9t33d3zmz9cgwgfiynhc HTTP/1.1
> Host: localhost:8065
> User-Agent: curl/7.71.1
> Accept: */*
> Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw
```

##### Response
```
< HTTP/1.1 200 OK
< Content-Type: application/json
< Expires: 0
< Vary: Accept-Encoding
< X-Request-Id: uku6s8nb8fn8xg7zhcf1urmwxr
< X-Version-Id: 5.28.0..ea71bf8023fe529a2125d71e6f29fa4b.true
< Date: Mon, 17 Aug 2020 12:01:12 GMT
< Content-Length: 204
```

##### Response body:
```json
{"id":"t8edyq9t33d3zmz9cgwgfiynhc","create_at":1597665637367,"user_id":"fc6suoon9pbbpmhrb9c967paxe","channel_id":"4i6jn8r483nnuqnibnmgz8jo4o","filename":"upload","file_size":7082489,"file_offset":6291456}
```

#### 4. Resume upload from the offset
```
dd if=upload skip=6291456 bs=1 | curl -v 'http://localhost:8065/api/v4/uploads/t8edyq9t33d3zmz9cgwgfiynhc' -H 'Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw' --data-binary @upload
```

##### Request
```
> POST /api/v4/uploads/t8edyq9t33d3zmz9cgwgfiynhc HTTP/1.1
> Host: localhost:8065
> User-Agent: curl/7.71.1
> Accept: */*
> Authorization: Bearer 9kg8nqrnxprd9jbykqeg4r51hw
> Content-Length: 791033
> Content-Type: application/x-www-form-urlencoded
```

##### Response
```
< HTTP/1.1 200 OK
< Content-Type: application/json
< Vary: Accept-Encoding
< X-Request-Id: 3wh4kzxq4pfaxd94ssu3aur8ge
< X-Version-Id: 5.28.0..ea71bf8023fe529a2125d71e6f29fa4b.true
< Date: Mon, 17 Aug 2020 12:01:48 GMT
< Content-Length: 201
```

#### Response body
```json
{"id":"g66jqep5npbsdjjc67orqzexnw","user_id":"fc6suoon9pbbpmhrb9c967paxe","create_at":1597665708465,"update_at":1597665708465,"delete_at":0,"name":"upload","extension":"","size":7082489,"mime_type":""}
```

**Comment by enahum** [comment]
@streamer45 this is perfect... I'm going to add that to the ticket, Thanks so much for taking the time to do that

**Comment by enahum** [comment]
@streamer45 one more thing in your example I see that in the request we have `Content-Type: application/x-www-form-urlencoded`, does this endpoint support `Content-Type: application/octet-stream` if not we'll need to include support for it.

**Comment by streamer45** [comment]
> @streamer45 one more thing in your example I see that in the request we have `Content-Type: application/x-www-form-urlencoded`, does this endpoint support `Content-Type: application/octet-stream` if not we'll need to include support for it.

The endpoint does not enforce any particular `content-type` so `application/octet-stream` should equally work.

**Comment by esethna** [comment]
@enahum do we have a ticket for the related mobile implementation side of this?

**Comment by enahum** [comment]
@esethna yes in v2

File: app/upload.go
Comment (streamer45) [reply app/upload.go:102]: @mgdelacroix 

The only problem I can see with this is that we end up making an additional store call to reset the session which could fail as well (since the original failed it's quite possible) and cause the same issue. I am not sure it's going to be easy to completely avoid all possible inconsistencies with the current design.

This could be potentially improved by *truncating* the stored file instead of removing everything. That way it's like the update never happened but it involves a bit more work on the `FileStore` side.

**Review by isacikgoz — APPROVED** [review]

**Comment by streamer45** [comment]
@enahum 

Just checking to see if you are okay with the proposed changes so I can move this forward.

**Comment by enahum** [comment]
@streamer45 code specific or flow?

If the flow is how you specified in your curl example, that looks ok to me.

If you are waiting for me to review the code, maybe ask someone else as my golang fu is extremely rusty by now

**Comment by streamer45** [comment]
> @streamer45 code specific or flow?
> 
> If the flow is how you specified in your curl example, that looks ok to me.
> 
> If you are waiting for me to review the code, maybe ask someone else as my golang fu is extremely rusty by now

Yeah, overall design is what I meant. Code review is already taken care of.

**Comment by enahum** [comment]
@streamer45 yes I’m ok with the design, very common practice and the previous endpoints remain. Go go go

File: app/upload.go
Comment (mgdelacroix) [reply app/upload.go:102]: Good point, I totally agree. I'd say that we can move forward with the current implementation and if it ever causes problems implement the file truncation in the store, I don't think this is a blocker

**Comment by streamer45** [comment]
@lindalumitchell 

this is a *small* new feature introducing a separate API endpoint for resumable uploads. It's purposely not altering the existing uploads/attachments functionality.
Not sure about the required QA process. I'd like to merge it soon so we can start build more stuff on top of this.

**Review by lindalumitchell — APPROVED** [review]
Thanks @streamer45 I read through the comments, and see tests in there as well, and I think we're good to approve this one through.

File: app/upload.go
Comment (jupenur) [inline app/upload.go:101]: If a user calls this multiple times in parallel by precisely timing their HTTP requests, wouldn't that allow them to upload arbitrarily large files?

1. User creates an upload session with file size 1000
2. User issues a request to submit 1 byte of data and transmits that 1 byte
  2.1. `us.FileOffset` is now 1
3. User issues another request to submit 1 byte of data, but pauses the request before actually sending any data over
  3.1. `us.FileOffset` doesn't get updated yet, remains at 1
4. User issues a request to submit 998 bytes and transmits all of the data
  4.1. `us.FileOffset` is now 1+998 = 999
5. User unpauses the earlier request and transmits the 1 byte of data
  5.1. `us.FileOffset` is updated again: 1+1 = 2

Repeat from step 3 until the offset reaches 1000, and the actual file size ends up being:

image

**Review by jupenur — CHANGES_REQUESTED** [review]

File: app/upload.go
Comment (streamer45) [reply app/upload.go:101]: That sounds correct. Concurrent requests can indeed cause some consistency issues including the one you suggested.
One option, and probably the simplest to implement, could be to allow only one upload request per upload session at any given time.

Comment (jupenur) [reply app/upload.go:101]: Yeah that sounds like the right approach.

File: app/upload.go
Comment (jupenur) [inline app/upload.go:111]: Not security-related, but this can cause cryptic error messages when `us.FileSize` is smaller than `minFirstPartSize`.

```suggestion
		if written < minFirstPartSize && written != us.FileSize && us.FileSize < minFirstPartSize {
```

**Review by jupenur — APPROVED** [review]

File: app/upload.go
Comment (streamer45) [reply app/upload.go:111]: The thing here is that we cannot resume (and should return error) uploads whose first segment/part is less than `minFirstPartSize` but we still want to support uploading files smaller than `minFirstPartSize` if they are done all at once (no resume), hence the need for an equality comparison.

The limited reader should already guarantee that `written` is never greater than `us.FileSize` and otherwise fail.

Comment (jupenur) [reply app/upload.go:111]: I think you misread the suggested change

Comment (jupenur) [reply app/upload.go:111]: Trying to clarify a bit; The current code in this PR returns the "first part too small" error in the case where `WriteFile` fails and writes nothing, including when the entire file is smaller than the minimum part size. It should return the real error instead, because the "too small" error doesn't make sense in that case.

By checking for `us.FileSize > minFirstPartSize` instead of `written != us.FileSize` we're excluding the small file case from this code path, but `written < minFirstPartSize && us.FileSize > minFirstPartSize` also implies `written != us.FileSize`.

Comment (jupenur) [reply app/upload.go:111]: Handling the error first sounds good to me.

Comment (streamer45) [reply app/upload.go:114]: It should be improved now. 
Let me know if that works and thanks for the precious feedback on this :+1:

Comment (jupenur) [reply app/upload.go:114]: Yeah looks good to me

**Pull request merged**
